### PR TITLE
SMS Log Report Optimization

### PR DESCRIPTION
**Pull request by esoergel**
This is hidden behind a "testing" feature flag.

This is an optimization allowing this change https://github.com/dimagi/commcare-hq/pull/23982 and in particular this change https://github.com/dimagi/commcare-hq/pull/24034 to perform on a larger scale.

The problem concerns how to get SMS event data - for many SMSs, this can be obtained by a double join: SMS -> MessagingSubevent -> Event.  In some cases, however, this must go SMS -> SQLXFormsSession -> MessagingSubEvent -> MessagingEvent, and not all of those connections are simple foreign keys.  My naive, POC implementation was a db hit per row to fetch that lookup.

@millerdev agreed to assist on this - he wrote a [CTE query](https://github.com/dimagi/commcare-hq/pull/24199/commits/71f427b77540ed721b60549f2400cc7b9ed32782) to do that second type of lookup.  I also came across [a cool blog post](https://medium.com/kolonial-no-product-tech/pushing-the-orm-to-its-limits-d26d87a66d28) describing the use of subqueries in a similar context, so [I tried that approach as well](https://github.com/dimagi/commcare-hq/pull/24199/commits/eb4eab00ad01d5cd44847b91e76089df09af8c01).  Both of those returned valid results, though they were still somewhat slow.  Pulling 4 months of data on a large project (500k rows) took about 22 seconds with the CTE query and 13 seconds with subqueries.  The baselines were 4 seconds doing a relational lookup only (incomplete event data), and a little over a second with no event data.

We then decided on a two-query approach - one for SMSs with subevents, and a second for those without.  Then the related fields are pulled in only on a smaller set of data.  With additional filtering on the `WITH` clause of the CTE query, it outperformed the subquery approach on the smaller dataset, so that's the approach I used going forward.  The two separate queries are then joined, with sorting preserved.  The standard library's `heapq.merge` does this beautifully, but it doesn't support `key` or `reverse` parameters until python 3, so I made my own implementation.

Overall, it feels like I might've gone a bit too far with the optimizations on this one, though I'm not sure what the best stopping point would've been.

### CODE REVIEW

File: corehq/apps/reports/standard/sms.py [hunk 1]
```diff
@@ -443,7 +445,19 @@ def order_by_col(data_):
...
         queryset = filter_by_location(queryset)
         queryset = order_by_col(queryset)
-        return queryset.select_related('messaging_subevent', 'messaging_subevent__parent')
+        return queryset
+
+    @property
+    def _sort_column(self):
+        col_fields = ['date', 'couch_recipient', 'phone_number', 'direction', 'text']
+        sort_col = self.request_params.get('iSortCol_0')
+        if sort_col is None or sort_col < len(col_fields):
```
Comment (millerdev) [inline corehq/apps/reports/standard/sms.py]:
```suggestion
        if sort_col is None or sort_col < 0 or sort_col >= len(col_fields):
```

**Review by millerdev — COMMENTED** [review]
I assume you got the query timing down to something more reasonable using the split/merge technique. We can discuss more offline, although some comments about your profiling findings and query implementation choices might be a nice addition in this PR for future reference.

File: corehq/util/itertools.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+from __future__ import unicode_literals
+
+from operator import lt, gt
+
+
+def next_or_Ellipsis(generator):
+    try:
+        return next(generator)
+    except StopIteration:
+        return Ellipsis
```
[the remark this answers is not in the document]
Comment (esoergel) [reply corehq/util/itertools.py]: That's a good point - I'll replace it with a locally defined constant.

> Also, the built-in next() function has a default argument

Neat, I didn't know about that

File: corehq/util/itertools.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+from operator import lt, gt
+
+
+def merge(generator1, generator2, key=None, reverse=False):
```
Comment (millerdev) [inline corehq/util/itertools.py]: Looks like this is not used anymore, so it could be removed (although it's a cool implementation).

**Review by millerdev — APPROVED** [review]

File: corehq/util/itertools.py (hunk 3 above)
Comment (esoergel) [reply corehq/util/itertools.py]: :disappointed: 
https://github.com/dimagi/commcare-hq/pull/24240

**Pull request merged**